### PR TITLE
GE Bug Fixes

### PR DESCRIPTION
**Pull request by enjeyw**
- rounding exchange rates amounts on sms's to 3 sig figs
- not requesting pin for exchange rate
- devious bug resulting in random pending transactions that stop transfer limits from working

### CODE REVIEW

File: app/server/models/utils.py [hunk 1]
```diff
@@ -14,6 +14,24 @@
 from server.exceptions import OrganisationNotProvidedException
 
 
+@contextmanager
+def ephemeral_alchemy_object(mod: db.Model, *args, **kwargs):
+    # weird SQLAlchemy behaviour cause object  to be persisted under some circumstances, even if they're not committed
+    # See: https://hades.github.io/2013/06/sqlalchemy-adds-objects-collections-automatically/
```
Comment (enjeyw) [inline app/server/models/utils.py]: this was so deviously evil

Comment (enjeyw) [reply app/server/models/utils.py]: @michielderoos hopefully this fixes it.

File: app/server/utils/ussd/token_processor.py [hunk 2]
```diff
@@ -73,61 +74,53 @@ def get_default_limit(user: User, token: Token, transfer_account: TransferAccoun
         :param transfer_account:
         :return: lowest limit applicable for a given CreditTransfer
         """
```
Comment (enjeyw) [inline app/server/utils/ussd/token_processor.py]: this is literally the same function, just wrapped in the ephemeral object context manager.

File: app/server/models/utils.py (hunk 1 above)
Comment (tristanhcole) [reply app/server/models/utils.py]: sounds weird as

File: app/server/utils/ussd/token_processor.py (hunk 2 above)
Comment (tristanhcole) [reply app/server/utils/ussd/token_processor.py]: so a credit transfer was being committed to the DB somehow?

File: app/server/utils/ussd/token_processor.py [hunk 3]
```diff
@@ -73,61 +74,53 @@ def get_default_limit(user: User, token: Token, transfer_account: TransferAccoun
...
+                        fraction = ge_limit[0].transfer_balance_fraction or 0
+                        if amount_avail < (fraction * transfer_account.balance)\
+                                and amount_avail < lowest_amount_avail:
+                            lowest_limit = limit
+                            lowest_amount_avail = amount_avail
+
+                if lowest_limit:
+                    return lowest_limit
+                else:
+                    return ge_limit[0]
```
Comment (tristanhcole) [inline app/server/utils/ussd/token_processor.py]: what's happened here @enjeyw ?

File: app/server/utils/ussd/token_processor.py [hunk 4]
```diff
@@ -73,61 +74,53 @@ def get_default_limit(user: User, token: Token, transfer_account: TransferAccoun
...
         if limit is not None and limit.transfer_balance_fraction is not None:
-            return round_amount(
-                limit.transfer_balance_fraction * TokenProcessor.get_balance(user)
-            )
+            return limit.transfer_balance_fraction * TokenProcessor.get_balance(user)
         elif limit.total_amount is not None:
-            return round_amount(
-                limit.total_amount
-            )
+            return limit.total_amount
```
Comment (tristanhcole) [inline app/server/utils/ussd/token_processor.py]: not rounding the amount anymore?

Comment (tristanhcole) [reply app/server/utils/ussd/token_processor.py]: ahhh you've moved to line 250... ok 👍

File: app/server/utils/ussd/token_processor.py (hunk 2 above)
Comment (enjeyw) [reply app/server/utils/ussd/token_processor.py]: yep! undocumented SQLAlchemy behaviour

Comment (tristanhcole) [reply app/server/utils/ussd/token_processor.py]: 👻

File: app/server/utils/ussd/token_processor.py (hunk 3 above)
Comment (enjeyw) [reply app/server/utils/ussd/token_processor.py]: nothing. That's my point above - the function is identical, the wrapping has just changed

File: app/server/utils/ussd/token_processor.py [hunk 5]
```diff
@@ -317,23 +314,7 @@ def send_token(sender: User, recipient: User, amount: float, reason_str: str, re
...
-
         try:
-            example_transfer.check_sender_transfer_limits()
-
-            try:
-                db.session.delete(example_transfer)
-            except InvalidRequestError:
-                # Instance is already deleted
-                pass
-
```
Comment (tristanhcole) [inline app/server/utils/ussd/token_processor.py]: this is being deprecated?

File: app/server/utils/ussd/token_processor.py (hunk 3 above)
Comment (tristanhcole) [reply app/server/utils/ussd/token_processor.py]: cool

**Pull request merged**